src/farming/library.cairo

```diff
@@ -0,0 +1,965 @@
+        let (len, cumsales) = Math.cumsum(len=len, xs=sales);
+        return (
+            len=len, times=times, prices=prices, updated_prices=updated_prices, cumsales=cumsales
+        );
+    }
+
+    func _compute_sales_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        index: felt,
+        absorption: felt,
+        negative_delta: felt,
+        positive_delta: felt,
+        len: felt,
+        times: felt*,
+        prices: felt*,
+        contract_address: felt,
+        slot: Uint256,
+        sales: felt*,
+        updated_prices: felt*,
+    ) -> (len: felt, updated_prices: felt*, sales: felt*) {
+        alloc_locals;
+
+        // [Check] End criteria
+        if (index == len) {
+            return (len=len, updated_prices=updated_prices, sales=sales);
+        }
+
+        // [Check] First iteration (special case)
+        let time = times[index];
+        let price = prices[index];
+        if (index == 0) {
+            // [Compute] Interpolated absorptions
+            let (first_time) = ICarbonableProject.getStartTime(
+                contract_address=contract_address, slot=slot
+            );
+            let (initial_absorption) = ICarbonableProject.getAbsorption(
+                contract_address=contract_address, slot=slot, time=first_time
+            );
+            let (final_absorption) = ICarbonableProject.getAbsorption(
+                contract_address=contract_address, slot=slot, time=time
+            );
+            let absorption = final_absorption - initial_absorption;
+            assert sales[index] = absorption * price;
+            assert updated_prices[index] = price;
+            return _compute_sales_iter(
+                index=index + 1,
+                absorption=absorption,
+                negative_delta=negative_delta,
+                positive_delta=positive_delta,
+                len=len,
+                times=times,
+                prices=prices,
+                contract_address=contract_address,
+                slot=slot,
+                sales=sales,
+                updated_prices=updated_prices,
+            );
+        }
+
+        // [Compute] Current time and price
+        let previous_time = times[index - 1];
+
+        // [Compute] Interpolated absorptions
+        let (initial_absorption) = ICarbonableProject.getAbsorption(
+            contract_address=contract_address, slot=slot, time=previous_time
+        );
+        let (final_absorption) = ICarbonableProject.getAbsorption(
+            contract_address=contract_address, slot=slot, time=time
+        );
+        let new_absorption = final_absorption - initial_absorption;
+
+        // [Check] Absorption cannot wrap a null absorption
+        with_attr error_message("CarbonableFarming: Prices cannot wrap a null absorption") {
+            assert_not_zero(new_absorption);
+        }
+
+        // [Compute] Delta to compensate from the previous sale
+        let real_sale = absorption * price;
```

Comment:
Maybe add some asserts to make sure price is within some range (not here but when it is set)
Otherwise ADMIN mistakes can rekt our contract state.